Tests/BackendIntegrationTests/OfflineStoreKitIntegrationTests.swift

```diff
@@ -0,0 +1,276 @@
+
+        Purchases.shared.invalidateCustomerInfoCache()
+        self.serverDown()
+
+        let info = try await Purchases.shared.customerInfo()
+        expect(info.entitlements.all).toNot(beEmpty())
+        try await self.verifyEntitlementWentThrough(info)
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testPurchaseWhileServerIsDownSucceedsButDoesNotFinishTransaction() async throws {
+        let logger = TestLogHandler()
+
+        self.serverDown()
+        try await self.purchaseMonthlyProduct()
+
+        logger.verifyMessageWasLogged(Strings.offlineEntitlements.computing_offline_customer_info, level: .info)
+        logger.verifyMessageWasNotLogged("Finishing transaction")
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testPurchaseWhileServerIsDownPostsReceiptAfterServerComesBack() async throws {
+        let logger = TestLogHandler()
+
+        // 1. Purchase while server is down
+        self.serverDown()
+        try await self.purchaseMonthlyProduct()
+
+        logger.verifyMessageWasNotLogged("Finishing transaction")
+
+        // 2. "Re-open" the app after the server is back
+        self.serverUp()
+        Purchases.shared.invalidateCustomerInfoCache()
+        await self.resetSingleton()
+
+        // 3. Ensure delegate is notified of subscription
+        try await asyncWait(
+            until: { [delegate = self.purchasesDelegate] in
+                delegate?.customerInfo?.activeSubscriptions.isEmpty == false
+            },
+            timeout: .seconds(5),
+            pollInterval: .milliseconds(200),
+            description: "Subscription never became active"
+        )
+
+        // 4. Ensure transaction is eventually finished
+        try await logger.verifyMessageIsEventuallyLogged(
+            "Finishing transaction",
+            level: .info,
+            timeout: .seconds(5),
+            pollInterval: .milliseconds(100)
+        )
+
+        // 5. Restart app again
+        Purchases.shared.invalidateCustomerInfoCache()
+        await self.resetSingleton()
+
+        // 6. To ensure (with a clean cache) that the receipt was posted
+        let info = try await Purchases.shared.customerInfo()
+        try await self.verifyEntitlementWentThrough(info)
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testReopeningAppWithOfflineEntitlementsDoesNotReturnStaleCache() async throws {
+        // 1. Purchase while server is down
+        self.serverDown()
+        try await self.purchaseMonthlyProduct()
+
+        // 2. "Re-open" the app
+        await self.resetSingleton()
+
+        // 3. `CustomerInfo` should contain offline purchase
+        let info = try await Purchases.shared.customerInfo()
+        try await self.verifyEntitlementWentThrough(info)
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testPurchaseAgainAfterServerRecovers() async throws {
```

Comment:
I believe apple/Google shouldn't allow you to purchase the same subscription a second time right? So not sure if this is a valid scenario.